### PR TITLE
Add support for approximate k-NN

### PR DESCRIPTION
**Pull request by sky-2002**
As mentioned in [this](https://github.com/pyg-team/pytorch_geometric/issues/7420) issue, I am adding code to support approximate nearest neighbour graphs.

### CODE REVIEW

File: torch_geometric/nn/pool/approx_knn.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# This file has code to create a KNN graph with approximate
+# nearest neighbors. We can call it approx-knn.
+
+import torch
+from pynndescent import NNDescent
```
Comment (rusty1s) [inline torch_geometric/nn/pool/approx_knn.py:5]: This likely needs to be made an optional dependency. You can achieve this by moving the import into the respective function.

File: torch_geometric/nn/pool/approx_knn.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def approx_nn(x, y, k, batch_x=None, batch_y=None):
```
Comment (rusty1s) [inline torch_geometric/nn/pool/approx_knn.py:8]: Can we add type hints here?

File: torch_geometric/nn/pool/approx_knn.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    x, y, = x / max_xy, y / max_xy
+
+    # Concat batch/features to ensure no cross-links between examples exist.
+    x = torch.cat([x, 2 * x.size(1) * batch_x.view(-1, 1).to(x.dtype)], dim=-1)
+    y = torch.cat([y, 2 * y.size(1) * batch_y.view(-1, 1).to(y.dtype)], dim=-1)
+
+    # Default metric is euclidean
+    # More can be found in the
+    # documentation here - https://pynndescent.readthedocs.io/en/latest/
+    index = NNDescent(x.detach().numpy())
```
Comment (rusty1s) [inline torch_geometric/nn/pool/approx_knn.py:37]: Missing `cpu()` call?

File: torch_geometric/nn/pool/approx_knn.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    max_xy = max(x.max().item(), y.max().item())
+    x, y, = x / max_xy, y / max_xy
+
+    # Concat batch/features to ensure no cross-links between examples exist.
+    x = torch.cat([x, 2 * x.size(1) * batch_x.view(-1, 1).to(x.dtype)], dim=-1)
+    y = torch.cat([y, 2 * y.size(1) * batch_y.view(-1, 1).to(y.dtype)], dim=-1)
+
+    # Default metric is euclidean
+    # More can be found in the
+    # documentation here - https://pynndescent.readthedocs.io/en/latest/
```
Comment (rusty1s) [inline torch_geometric/nn/pool/approx_knn.py:36]: I think this scheme here only works for Euclidean due to how we integrate `batch` into `x`/`y`.

**Review by rusty1s — COMMENTED** [review]
Thanks for this PR. Added some comments. The next steps would be
* Add a line to `CHANGELOG.md` that describes the feature you add here
* Add a corresponding test in `test/nn/pool/test_approx_knn.py`. This can be similar to https://github.com/rusty1s/pytorch_cluster/blob/master/test/test_knn.py#L14-L48. Note that you likely need to add the `@withPackage('pynndescent')` decorator to let CI pass.

File: torch_geometric/nn/pool/approx_knn.py (hunk 1 above)
Comment (sky-2002) [reply torch_geometric/nn/pool/approx_knn.py:5]: Done

File: torch_geometric/nn/pool/approx_knn.py (hunk 4 above)
Comment (sky-2002) [reply torch_geometric/nn/pool/approx_knn.py:36]: ok, can you elaborate?

**Review by rusty1s — APPROVED** [review]

**Pull request merged**